### ISSUE TITLE
implement the KHR_draco_mesh_compression extension

### ISSUE THREAD

**Issue by stephengold** [jMonkeyEngine/jmonkeyengine#2118]
Khronos has specified [an extension to glTF](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_draco_mesh_compression/README.md) for loading models that employ [Draco mesh compression](https://google.github.io/draco/).

Currently, attempting to load such a model results in warnings and an empty result:

```console
Oct 17, 2023 11:50:50 AM com.jme3.scene.plugins.gltf.CustomContentManager readExtension
WARNING: Could not find loader for extension KHR_draco_mesh_compression
```

Since the extension may be widely used in the future, it would be nice if jme3-plugins could handle it.

**Comment by javagl** [comment]
There are not too many options for decoding Draco in Java. 

I tried using the official WASM distribution of Draco with GraalWASM, but ... that doesn't seem like a promising path right now. 

There's always the option to use JNI, and that could be a reasonable option, but always carries some overhead for compiling and deploying the native libraries. (Maybe LWJGL will jump in here. They already created bindings for [Meshoptimizer](https://javadoc.lwjgl.org/org/lwjgl/util/meshoptimizer/MeshOptimizer.html), meaning that support for [`EXT_meshopt_compression`](https://github.com/KhronosGroup/glTF/tree/main/extensions/2.0/Vendor/EXT_meshopt_compression) may be in reach, and maybe even relatively easy... ).

In the meantime: I had some success with decoding Draco-compressed data from glTF with https://github.com/fileformat-drako/FileFormat.Drako-for-Java . I can not yet strongly "reommend" that library, because it's relatively new, and I can not yet say much about stability and performance. But it's promising, and I'll continue to investigate it. Maybe someone here also wants to have a look.

**Comment by stephengold** [comment]
At the present time, I believe our best solution is to use [MonkeyWrench](https://github.com/stephengold/MonkeyWrench). Of course, as its author I may be biased.

**Comment by javagl** [comment]
@stephengold From _quickly_ skimming over that, it looks like it is using Assimp for the actual import/loading of the glTF asset, which, in turn uses the [LWJGL Assimp bindings](https://javadoc.lwjgl.org/org/lwjgl/assimp/Assimp.html). 

(Iff this is correct, then it already _does_ boil down to LWJGL...)

I don't know the details about how well Assimp itself supports glTF. I think I heard some rumours that the import may ~"have some issues", but I don't know how true that is or how severe these issues are. The [Fileformts.md](https://github.com/assimp/assimp/blob/master/doc/Fileformats.md) lists glTF 2.0 with a few extensions, where Draco is **not** included. When doing a quick code search for `"draco"`, then it seems to appear in the code in the context of glTF. So I don't know what the truth is.

In any case: If MonkeyWrench->(LWJGL)Assimp "works" for glTF+Draco, then this may very well be the most reasonable path here. And the fact that the Draco decoding itself would then be "hidden" behind the Assimp API may, in fact, be a plus: When something's wrong and you have to choice of "Fixing it" or "Opening an issue in another repo", the latter tends to be easier...

**Comment by stephengold** [comment]
Yes, MonkeyWrench boils down to LWJGL and Assimp. I believe Assimp supports Draco because I tested it on Draco models found in  KhronosGroup/glTF-Sample-Assets and KhronosGroup/glTF-Sample-Models.

However, depending on MonkeyWrench for our Draco support isn't ideal because of the 2-year lag for getting Assimp bugfixes into MonkeyWrench. I'd be much happier if jme3-plugins also supported Draco.

**Comment by riccardobl** [comment]
I think there is a java implementation now https://products.openize.com/drako/java/

**Comment by javagl** [comment]
@riccardobl Yes, that's the https://github.com/openize-com/openize-drako-java that I already linked to above. I tried it out for a few input data sets, but not "deeply, properly" integrated into any form of glTF loader. Somewhat recently, someone asked about how to use that library in the context of glTF, at https://github.com/openize-com/openize-drako-java/issues/4 , and I added a small comment and code snippet there as well - maybe a starter for people who want to try it out.

**Comment by stephengold** [comment]
I promise not to work on this issue for at least a month. If someone wants the bounty, there's no need to wait.

**Comment by javagl** [comment]
I actually already started to ""work"" on this. But this started at 0 for now - cloning jMonkeyEngine, trying to set things up (never having used Gradle before) ... Now... I can start `TestGltfLoading`, and trying to load a draco-compressed asset already prints some
`System.out.println("Trying to handle draco...");`
that I inserted into some (otherwise empty) `DracoMeshCompressionExtensionLoader`. Maybe it's not sooo hard to fill this with life, but 1. I cannot make any promises or "strong" committment to actually _doing_ (i.e. 'resolving') this, and 2. I don't want to "block" anyone from grabbing the bounty. First come, first serve, I guess 🤷

**Comment by riccardobl** [comment]
Don't worry, things move slow in our project, anyway. 
Feel free to ping me here, on our [forum](https://hub.jmonkeyengine.org/) or even on [discord](https://discord.gg/Rv8sfVu) if you need any insight in the engine internals.

I would prefer to assign this to you for the time being, is that ok? Or do you prefer to leave it unassigned?

**Comment by javagl** [comment]
It's OK to assign me (and be it only as a reminder...). 

EDIT: See update below <strike>But if someone comments here, saying that he wants to do it (and grab the bounty), the assignment could be changed - I don't want to claim the "exclusive right" of looking into this.</strike>

**Comment by javagl** [comment]
I revoke the statement that I won't claim that "exclusive right". Maybe I can grab that bounty with reasonable effort. At least, this is the Draco-compressed version of the "BoomBox" sample asset...

<img width="668" height="559" alt="Image" src="https://github.com/user-attachments/assets/cfb2a610-27b2-4195-8caa-60bf554a9b36" />

I'll clean it up and open a draft PR soon. There are some pretty fundamental things that I don't know - how the heck do you (properly) add dependencies with Gradle? What should the general error handling strategy be? Etc. But maybe we can iterate on that PR as necessary.

**Comment by javagl** [comment]
Well, it's never as easy as it looks. 

In order to test all this, I wanted to make sure that all Draco-encoded models from https://github.com/KhronosGroup/glTF-Sample-Assets can be decoded properly (as some sort of baseline test). This would be the following ones:
```java
        String dracoTestModel = null;
        //dracoTestModel = "Avocado";
        //dracoTestModel = "BarramundiFish";
        //dracoTestModel = "BoomBox";
        //dracoTestModel = "BrainStem";
        //dracoTestModel = "CesiumMan";
        //dracoTestModel = "CesiumMilkTruck";
        //dracoTestModel = "Corset";
        //dracoTestModel = "Lantern";
        //dracoTestModel = "MorphPrimitivesTest";
        //dracoTestModel = "RiggedFigure";
        //dracoTestModel = "RiggedSimple";
        //dracoTestModel = "SunglassesKhronos";
        //dracoTestModel = "VirtualCity";
        //dracoTestModel = "WaterBottle";

``` 
(Box and Duck omitted)

---

There are several issues for now:


---

Models that contain vertex skinning cannot be decoded properly. The `JOINTS_0` data cannot be read. I already zoomed into that, and am reasonably sure that this is an issue/limitation of `openize-drako-java`. I opened https://github.com/openize-com/openize-drako-java/issues/5 with a detailed description and test data to reproduce it.

(I'm a bit concerned: The repository does not seem to be sooo actively maintained - let's hope one of the develpers finds the time to take a look and chime in there...)

---

The issue also mentions another case where the decoding fails. This is unrelated to the vertex skinning. It _might_ be something that is related to the Draco bitstream version (maybe a few feature that is not yet supported?). Also something that will have to be investigated further, hopefully the developers can sort this out.

Fortunately, this only affected _one_ of the models for now, namely the `SunglassesKhronos`

---

Something is off in the cases where the draco-encoded meshes are attached to nodes with translation:

<img width="636" height="509" alt="Image" src="https://github.com/user-attachments/assets/82eaefaa-a6d4-44ee-a10e-f08bf683d244" />

I have no idea where this is coming from. It might be related to the quantization in some way, but that remains to be investigated.


---

Probably low priority: I *think* that the Draco version of one of the glTF Sample Assets is actually invalid. Unless I'm overlooking something in the spec, there are some inconsistencies between some `accessor.count` values and the values that are stored as the indices for accessing them.

A description of that is in https://github.com/KhronosGroup/glTF-Sample-Assets/issues/264

---

But ... yeah. _Some_ models are working. I'll still try to clean it up and create a draft PR...

**Comment by stephengold** [comment]
JMonkeyEngine's glTF loader has many other defects/issues besides this one. I believe we can close this issue even if some of the 16 sample assets with Draco and some of the 19 sample models with Draco don't load perfectly. It would be particularly helpful to compare similar models with and without Draco. I'll leave the final judgment to others who are currently more active than I am.

**Comment by riccardobl** [comment]
I agree. We'd be happy to review a PR that adds KHR_draco_mesh_compression support to the loader even if the underlying implementation has some defects.

We can address the limitations/issues in openize-drako separately on our side.
Thanks for reporting them, it’s helpful to know where the decoder falls short so we know what to investigate next.

**Comment by javagl** [comment]
> It would be particularly helpful to compare similar models with and without Draco

The lines below the list of test cases that I posted already are
```java
        boolean testDraco = false;
        if (testDraco)
        {
        	dracoTestFlavor = "glTF-Draco";
        	dracoTestExtension = "gltf";
        }
        else
        {
            dracoTestFlavor = "glTF-Binary";
            dracoTestExtension = "glb";
        }
        // Assemble a path like 
        // "Models/gltf/BoomBox/glTF-Draco/BoomBox.gltf"
        String dracoTestPath = "Models/gltf/" + dracoTestModel + "/" + dracoTestFlavor + "/" + dracoTestModel + "." + dracoTestExtension; 
```
to be able to quickly "switch" between the Draco- and non-Draco version 🙂 I see that as a form of "baseline comparison" for a task like this. 

This is particularly useful for things like vertex skinning. This _is_ a complicated topic (and usually raises the largest number of issues/questions - see the discussion and crosslinks at the bottom of https://github.com/KhronosGroup/glTF-Tutorials/pull/64 ...). I would have anticipated that some forms of skinning might have issues in jME. But (even though I have not yet cross-checked all examples) it looks like the skinning examples _generally_ work in jME. The further investigation then led me to believe that the issue is rooted in `openize-drako-java` itself.

I'll have another look at the issue with the translations (i.e. the wheels in the screenshot), because this affects "most" models (except for some of the showcase sample assets that happen to have no node translations). For things that are out of the control of jME (like the skinning), it will have to be tracked as a limitation/issue, with some degrees of freedom of how to handle that during loading (ignore/warn/throw) - details TBD in the PR.

**Comment by javagl** [comment]
> JMonkeyEngine's glTF loader has many other defects/issues besides this one.

I have the tendency to see ~"room for improvement" when diving into code, and find it hard to not **do** this improvement. This can be dangerous for a PR, because it can become harder to review, and (in the worst case) cause regressions. 

So I'll try to keep the scope of the upcoming PR narrow and focussed on the Draco part. 

But when I see that the loader currently checks for `.bin` file extensions at https://github.com/jMonkeyEngine/jmonkeyengine/blob/8793765c745134606ea88c57ea38b834ffd771ad/jme3-plugins/src/gltf/java/com/jme3/scene/plugins/gltf/GltfLoader.java#L653 and bails out for other extensions, then I think that this should be fixed. 

According to the specification at https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#file-extensions-and-media-types , the file extension **SHOULD** be `.bin` or `.glbin`. (I've never seen `.glbin`, but that doesn't matter). The glTF specification is _very_ precise, and _strictly_ obeys [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119), and "**SHOULD**" means that...

> ... there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

Here, this means that the loader should _not_ expect a particular file extension. 

(I'll probably not fix this here, but will try to collect other things where I see "room for improvement" - maybe they can be collected in an issue and handled with some ~"cleanup PR")

**Comment by riccardobl** [comment]
Sure, feel free to open as many new issues as you like with your findings. Even just pointing out the problem, without too much elaboration, is enough to help us keep track of what needs improvement.

If you prefer to report all the issues here, I can open dedicated issues myself.

**Comment by javagl** [comment]
I opened a first draft PR at https://github.com/jMonkeyEngine/jmonkeyengine/pull/2591 , maybe someone wants to have a look.

> Sure, feel free to open as many new issues as you like with your findings. Even just pointing out the problem, without too much elaboration, is enough to help us keep track of what needs improvement.
> 
> If you prefer to report all the issues here, I can open dedicated issues myself.

I'm not sure about the best process here. The `.bin`-check was just something that I stumbled over. I did not yet "thoroughly review" the `GltfLoader`, but will probably have a closer look, and then, the `.bin` thing may just be one of several "nitpicks", and then I'll see whether it could be an issue with a short list, or whether I'd immediately create a PR with suggested fixes.

**Comment by riccardobl** [comment]
I am relabelling this from `medium-hard` to `hard`, due to the sheer amount of work that was needed to debug, report and test the issues with the upstream library
